cloudmarker/clouds/gcpcloud.py

```diff
@@ -150,61 +154,23 @@ def read(self):
         Yields:
             dict: Firewall rule or VM instance configuration data.
 
-        Here is an example of a firewall rule record :obj:`dict` yielded
-        by this method:
-
-        .. code:: python
-
-            {
-              "kind": "compute#firewall",
-              "id": "7890789078907890",
-              "creationTimestamp": "2018-12-19T01:43:51.988-08:00",
-              "name": "default-allow-ss",
-              "description": "str",
-              "network": "https://www.googleapis.com/compute/v1/"
-                         "projects/foo/global/networks/default",
-              "priority": 1000,
-              "sourceRanges": [
-                "0.0.0.0/0"
-              ],
-              "targetTags": [
-                "https-server"
-              ],
-              "allowed": [
-                {
-                  "IPProtocol": "tcp",
-                  "ports": [
-                    "80",
-                    "8080-8090"
-                  ]
-                }
-              ],
-              "direction": "INGRESS",
-              "logConfig": {
-                "enable": bool
-              },
-              "disabled": bool,
-              "selfLink": "https://www.googleapis.com/compute/v1/"
-                          "projects/foo/global/firewalls/default-allow-https",
-              "record_type": "firewall_rule"
-            }
-
         """
-        firewall_rules = self._get_firewall_rules()
+        firewalls = self._get_firewalls()
         instances = self._get_instances()
 
-        for rule in firewall_rules:
+        for firewall in firewalls:
             record = {
-                'raw': rule,
+                'raw': firewall,
                 'ext': {
-                    'record_type': 'firewall_rule'
+                    'record_type': 'firewall'
                 },
                 'com': {
                     'cloud_type': 'gcp',
-                    'record_type': 'firewall_rule'
+                    'record_type': None,
                 }
             }
             yield record
```

Comment:
Why are you yielding the original record, even after the record is split in multiple records after normalization?

Reply:
The original record is yielded here so that it remains preserved in our data store. We do not have storage constraint right now. Most of the data stores supported by this project would be big data systems. Also, the original raw `firewall` record has more information than what these normalized `firewall_rule` records have. So I do not see a good reason to lose the original `firewall` data.

Also, here at [line 252](https://github.com/cloudmarker/cloudmarker/pull/65/files#diff-c403e664c38d6edc528237cc5e434e05R252), we are keeping a `reference` to the original raw record in the `firewall_rule` record. In the cloud-independent `FirewallRuleEvent` plugin in #69, this `reference` would be included in every event generated. Thus, when a user of this project (a security analyst for example) reads an alert, he/she can find the original raw `firewall` record using this `reference` if he/she wishes too.